<a href="https://colab.research.google.com/github/Olga13579/Data_Circle_Project-Stack_Overflow/blob/feature%2Floading-pershina/notebooks/01_loading_pershina.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
import pandas as pd
base = "https://github.com/StackExchange/Survey/raw/refs/heads/main/packages/archive/{}/results.csv"


In [4]:
df_2022 = pd.read_csv(base.format(2022))
print(df_2022.shape)
df_2022.head()

(73268, 79)


,ResponseId,MainBranch,Employment,RemoteWork,CodingActivities,EdLevel,LearnCode,LearnCodeOnline,LearnCodeCoursesCert,YearsCode,...,TimeSearching,TimeAnswering,Onboarding,ProfessionalTech,TrueFalse_1,TrueFalse_2,TrueFalse_3,SurveyLength,SurveyEase,ConvertedCompYearly
0,1,None of these,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2,I am a developer by profession,"Employed, full-time",Fully remote,Hobby;Contribute to open-source projects,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Too long,Difficult,NaN
2,3,"I am not primarily a developer, but I write co...","Employed, full-time","Hybrid (some remote, some in-person)",Hobby,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",Books / Physical media;Friend or family member...,Technical documentation;Blogs;Programming Game...,NaN,14,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Appropriate in length,Neither easy nor difficult,40205.0
3,4,I am a developer by profession,"Employed, full-time",Fully remote,I don’t code outside of work,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)","Books / Physical media;School (i.e., Universit...",NaN,NaN,20,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Appropriate in length,Easy,215232.0
4,5,I am a developer by profession,"Employed, full-time","Hybrid (some remote, some in-person)",Hobby,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)","Other online resources (e.g., videos, blogs, f...",Technical documentation;Blogs;Stack Overflow;O...,NaN,8,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Too long,Easy,NaN


In [5]:
df_2023 = pd.read_csv(base.format(2023))
print(df_2023.shape)
df_2023.head()

(89184, 84)


,ResponseId,Q120,MainBranch,Age,Employment,RemoteWork,CodingActivities,EdLevel,LearnCode,LearnCodeOnline,...,Frequency_1,Frequency_2,Frequency_3,TimeSearching,TimeAnswering,ProfessionalTech,Industry,SurveyLength,SurveyEase,ConvertedCompYearly
0,1,I agree,None of these,18-24 years old,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2,I agree,I am a developer by profession,25-34 years old,"Employed, full-time",Remote,Hobby;Contribute to open-source projects;Boots...,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)",Books / Physical media;Colleague;Friend or fam...,Formal documentation provided by the owner of ...,...,1-2 times a week,10+ times a week,Never,15-30 minutes a day,15-30 minutes a day,DevOps function;Microservices;Automated testin...,"Information Services, IT, Software Development...",Appropriate in length,Easy,285000.0
2,3,I agree,I am a developer by profession,45-54 years old,"Employed, full-time","Hybrid (some remote, some in-person)",Hobby;Professional development or self-paced l...,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)",Books / Physical media;Colleague;On the job tr...,Formal documentation provided by the owner of ...,...,6-10 times a week,6-10 times a week,3-5 times a week,30-60 minutes a day,30-60 minutes a day,DevOps function;Microservices;Automated testin...,"Information Services, IT, Software Development...",Appropriate in length,Easy,250000.0
3,4,I agree,I am a developer by profession,25-34 years old,"Employed, full-time","Hybrid (some remote, some in-person)",Hobby,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)",Colleague;Friend or family member;Other online...,Formal documentation provided by the owner of ...,...,1-2 times a week,10+ times a week,1-2 times a week,15-30 minutes a day,30-60 minutes a day,Automated testing;Continuous integration (CI) ...,NaN,Appropriate in length,Easy,156000.0
4,5,I agree,I am a developer by profession,25-34 years old,"Employed, full-time;Independent contractor, fr...",Remote,Hobby;Contribute to open-source projects;Profe...,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)",Books / Physical media;Online Courses or Certi...,Formal documentation provided by the owner of ...,...,1-2 times a week,1-2 times a week,3-5 times a week,60-120 minutes a day,30-60 minutes a day,Microservices;Automated testing;Observability ...,Other,Appropriate in length,Neither easy nor difficult,23456.0


In [6]:
df_2024 = pd.read_csv(base.format(2024))
print(df_2024.shape)
df_2024.head()

(65437, 114)


,ResponseId,MainBranch,Age,Employment,RemoteWork,Check,CodingActivities,EdLevel,LearnCode,LearnCodeOnline,...,JobSatPoints_6,JobSatPoints_7,JobSatPoints_8,JobSatPoints_9,JobSatPoints_10,JobSatPoints_11,SurveyLength,SurveyEase,ConvertedCompYearly,JobSat
0,1,I am a developer by profession,Under 18 years old,"Employed, full-time",Remote,Apples,Hobby,Primary/elementary school,Books / Physical media,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2,I am a developer by profession,35-44 years old,"Employed, full-time",Remote,Apples,Hobby;Contribute to open-source projects;Other...,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)",Books / Physical media;Colleague;On the job tr...,Technical documentation;Blogs;Books;Written Tu...,...,0.0,0.0,0.0,0.0,0.0,0.0,NaN,NaN,NaN,NaN
2,3,I am a developer by profession,45-54 years old,"Employed, full-time",Remote,Apples,Hobby;Contribute to open-source projects;Other...,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",Books / Physical media;Colleague;On the job tr...,Technical documentation;Blogs;Books;Written Tu...,...,NaN,NaN,NaN,NaN,NaN,NaN,Appropriate in length,Easy,NaN,NaN
3,4,I am learning to code,18-24 years old,"Student, full-time",NaN,Apples,NaN,Some college/university study without earning ...,"Other online resources (e.g., videos, blogs, f...",Stack Overflow;How-to videos;Interactive tutorial,...,NaN,NaN,NaN,NaN,NaN,NaN,Too long,Easy,NaN,NaN
4,5,I am a developer by profession,18-24 years old,"Student, full-time",NaN,Apples,NaN,"Secondary school (e.g. American high school, G...","Other online resources (e.g., videos, blogs, f...",Technical documentation;Blogs;Written Tutorial...,...,NaN,NaN,NaN,NaN,NaN,NaN,Too short,Easy,NaN,NaN


In [7]:
df_2025 = pd.read_csv(base.format(2025))
print(df_2025.shape)
df_2025.head()

/tmp/ipykernel_1692/1779289391.py:1: DtypeWarning: Columns (56,74,92,97,98,105,109,110,132,162,165) have mixed types. Specify dtype option on import or set low_memory=False.
  df_2025 = pd.read_csv(base.format(2025))


(49191, 172)


,ResponseId,MainBranch,Age,EdLevel,Employment,EmploymentAddl,WorkExp,LearnCodeChoose,LearnCode,LearnCodeAI,...,AIAgentOrchestration,AIAgentOrchWrite,AIAgentObserveSecure,AIAgentObsWrite,AIAgentExternal,AIAgentExtWrite,AIHuman,AIOpen,ConvertedCompYearly,JobSat
0,1,I am a developer by profession,25-34 years old,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",Employed,"Caring for dependents (children, elderly, etc.)",8.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",...,Vertex AI,NaN,NaN,NaN,ChatGPT,NaN,When I don’t trust AI’s answers,"Troubleshooting, profiling, debugging",61256.0,10.0
1,2,I am a developer by profession,25-34 years old,"Associate degree (A.A., A.S., etc.)",Employed,NaN,2.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,NaN,NaN,When I don’t trust AI’s answers;When I want to...,All skills. AI is a flop.,104413.0,9.0
2,3,I am a developer by profession,35-44 years old,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)","Independent contractor, freelancer, or self-em...",None of the above,10.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,ChatGPT;Claude Code;GitHub Copilot;Google Gemini,NaN,When I don’t trust AI’s answers;When I want to...,"Understand how things actually work, problem s...",53061.0,8.0
3,4,I am a developer by profession,35-44 years old,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)",Employed,None of the above,4.0,"Yes, I am not new to coding but am learning ne...","Other online resources (e.g. standard search, ...","Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,ChatGPT;Claude Code,NaN,When I don’t trust AI’s answers;When I want to...,NaN,36197.0,6.0
4,5,I am a developer by profession,35-44 years old,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)","Independent contractor, freelancer, or self-em...","Caring for dependents (children, elderly, etc.)",21.0,"No, I am not new to coding and did not learn n...",NaN,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,NaN,NaN,When I don’t trust AI’s answers,"critical thinking, the skill to define the tas...",60000.0,7.0


In [8]:
df_2022.columns

Index(['ResponseId', 'MainBranch', 'Employment', 'RemoteWork',
       'CodingActivities', 'EdLevel', 'LearnCode', 'LearnCodeOnline',
       'LearnCodeCoursesCert', 'YearsCode', 'YearsCodePro', 'DevType',
       'OrgSize', 'PurchaseInfluence', 'BuyNewTool', 'Country', 'Currency',
       'CompTotal', 'CompFreq', 'LanguageHaveWorkedWith',
       'LanguageWantToWorkWith', 'DatabaseHaveWorkedWith',
       'DatabaseWantToWorkWith', 'PlatformHaveWorkedWith',
       'PlatformWantToWorkWith', 'WebframeHaveWorkedWith',
       'WebframeWantToWorkWith', 'MiscTechHaveWorkedWith',
       'MiscTechWantToWorkWith', 'ToolsTechHaveWorkedWith',
       'ToolsTechWantToWorkWith', 'NEWCollabToolsHaveWorkedWith',
       'NEWCollabToolsWantToWorkWith', 'OpSysProfessional use',
       'OpSysPersonal use', 'VersionControlSystem', 'VCInteraction',
       'VCHostingPersonal use', 'VCHostingProfessional use',
       'OfficeStackAsyncHaveWorkedWith', 'OfficeStackAsyncWantToWorkWith',
       'OfficeStackSyncHaveWork

In [9]:
df_2023.columns

Index(['ResponseId', 'Q120', 'MainBranch', 'Age', 'Employment', 'RemoteWork',
       'CodingActivities', 'EdLevel', 'LearnCode', 'LearnCodeOnline',
       'LearnCodeCoursesCert', 'YearsCode', 'YearsCodePro', 'DevType',
       'OrgSize', 'PurchaseInfluence', 'TechList', 'BuyNewTool', 'Country',
       'Currency', 'CompTotal', 'LanguageHaveWorkedWith',
       'LanguageWantToWorkWith', 'DatabaseHaveWorkedWith',
       'DatabaseWantToWorkWith', 'PlatformHaveWorkedWith',
       'PlatformWantToWorkWith', 'WebframeHaveWorkedWith',
       'WebframeWantToWorkWith', 'MiscTechHaveWorkedWith',
       'MiscTechWantToWorkWith', 'ToolsTechHaveWorkedWith',
       'ToolsTechWantToWorkWith', 'NEWCollabToolsHaveWorkedWith',
       'NEWCollabToolsWantToWorkWith', 'OpSysPersonal use',
       'OpSysProfessional use', 'OfficeStackAsyncHaveWorkedWith',
       'OfficeStackAsyncWantToWorkWith', 'OfficeStackSyncHaveWorkedWith',
       'OfficeStackSyncWantToWorkWith', 'AISearchHaveWorkedWith',
       'AISearchWan

In [10]:
df_2024.columns

Index(['ResponseId', 'MainBranch', 'Age', 'Employment', 'RemoteWork', 'Check',
       'CodingActivities', 'EdLevel', 'LearnCode', 'LearnCodeOnline',
       ...
       'JobSatPoints_6', 'JobSatPoints_7', 'JobSatPoints_8', 'JobSatPoints_9',
       'JobSatPoints_10', 'JobSatPoints_11', 'SurveyLength', 'SurveyEase',
       'ConvertedCompYearly', 'JobSat'],
      dtype='object', length=114)

In [11]:
df_2025.columns

Index(['ResponseId', 'MainBranch', 'Age', 'EdLevel', 'Employment',
       'EmploymentAddl', 'WorkExp', 'LearnCodeChoose', 'LearnCode',
       'LearnCodeAI',
       ...
       'AIAgentOrchestration', 'AIAgentOrchWrite', 'AIAgentObserveSecure',
       'AIAgentObsWrite', 'AIAgentExternal', 'AIAgentExtWrite', 'AIHuman',
       'AIOpen', 'ConvertedCompYearly', 'JobSat'],
      dtype='object', length=172)

In [12]:
df_2022.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 73268 entries, 0 to 73267
Data columns (total 79 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   ResponseId                      73268 non-null  int64  
 1   MainBranch                      73268 non-null  object 
 2   Employment                      71709 non-null  object 
 3   RemoteWork                      58958 non-null  object 
 4   CodingActivities                58899 non-null  object 
 5   EdLevel                         71571 non-null  object 
 6   LearnCode                       71580 non-null  object 
 7   LearnCodeOnline                 50685 non-null  object 
 8   LearnCodeCoursesCert            29389 non-null  object 
 9   YearsCode                       71331 non-null  object 
 10  YearsCodePro                    51833 non-null  object 
 11  DevType                         61302 non-null  object 
 12  OrgSize                         

In [13]:
df_2023.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89184 entries, 0 to 89183
Data columns (total 84 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   ResponseId                           89184 non-null  int64  
 1   Q120                                 89184 non-null  object 
 2   MainBranch                           89184 non-null  object 
 3   Age                                  89184 non-null  object 
 4   Employment                           87898 non-null  object 
 5   RemoteWork                           73810 non-null  object 
 6   CodingActivities                     73764 non-null  object 
 7   EdLevel                              87973 non-null  object 
 8   LearnCode                            87663 non-null  object 
 9   LearnCodeOnline                      70084 non-null  object 
 10  LearnCodeCoursesCert                 37076 non-null  object 
 11  YearsCode                   

In [14]:
df_2024.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65437 entries, 0 to 65436
Data columns (total 114 columns):
 #    Column                          Non-Null Count  Dtype  
---   ------                          --------------  -----  
 0    ResponseId                      65437 non-null  int64  
 1    MainBranch                      65437 non-null  object 
 2    Age                             65437 non-null  object 
 3    Employment                      65437 non-null  object 
 4    RemoteWork                      54806 non-null  object 
 5    Check                           65437 non-null  object 
 6    CodingActivities                54466 non-null  object 
 7    EdLevel                         60784 non-null  object 
 8    LearnCode                       60488 non-null  object 
 9    LearnCodeOnline                 49237 non-null  object 
 10   TechDoc                         40897 non-null  object 
 11   YearsCode                       59869 non-null  object 
 12   YearsCodePro    

In [15]:
df_2025.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49191 entries, 0 to 49190
Data columns (total 172 columns):
 #    Column                                    Non-Null Count  Dtype  
---   ------                                    --------------  -----  
 0    ResponseId                                49191 non-null  int64  
 1    MainBranch                                49191 non-null  object 
 2    Age                                       49191 non-null  object 
 3    EdLevel                                   48149 non-null  object 
 4    Employment                                48339 non-null  object 
 5    EmploymentAddl                            44875 non-null  object 
 6    WorkExp                                   42893 non-null  float64
 7    LearnCodeChoose                           46858 non-null  object 
 8    LearnCode                                 33556 non-null  object 
 9    LearnCodeAI                               45201 non-null  object 
 10   AILearnHow          

In [16]:
cols_2022 = set(df_2022.columns)

In [17]:
cols_2023 = set(df_2023.columns)

In [18]:
cols_2024 = set(df_2024.columns)

In [19]:
cols_2025 = set(df_2025.columns)

In [20]:
common_cols = (cols_2022) & (cols_2023) & (cols_2024) & (cols_2025)
common_cols


{'Age',
 'CompTotal',
 'ConvertedCompYearly',
 'Country',
 'Currency',
 'DatabaseHaveWorkedWith',
 'DatabaseWantToWorkWith',
 'DevType',
 'EdLevel',
 'Employment',
 'ICorPM',
 'LanguageHaveWorkedWith',
 'LanguageWantToWorkWith',
 'LearnCode',
 'MainBranch',
 'OfficeStackAsyncHaveWorkedWith',
 'OfficeStackAsyncWantToWorkWith',
 'OpSysPersonal use',
 'OpSysProfessional use',
 'OrgSize',
 'PlatformHaveWorkedWith',
 'PlatformWantToWorkWith',
 'PurchaseInfluence',
 'RemoteWork',
 'ResponseId',
 'SOAccount',
 'SOComm',
 'SOPartFreq',
 'SOVisitFreq',
 'WebframeHaveWorkedWith',
 'WebframeWantToWorkWith',
 'WorkExp',
 'YearsCode'}

In [21]:
print(len(common_cols))

33


In [22]:
cols = sorted(common_cols)
print(cols)

['Age', 'CompTotal', 'ConvertedCompYearly', 'Country', 'Currency', 'DatabaseHaveWorkedWith', 'DatabaseWantToWorkWith', 'DevType', 'EdLevel', 'Employment', 'ICorPM', 'LanguageHaveWorkedWith', 'LanguageWantToWorkWith', 'LearnCode', 'MainBranch', 'OfficeStackAsyncHaveWorkedWith', 'OfficeStackAsyncWantToWorkWith', 'OpSysPersonal use', 'OpSysProfessional use', 'OrgSize', 'PlatformHaveWorkedWith', 'PlatformWantToWorkWith', 'PurchaseInfluence', 'RemoteWork', 'ResponseId', 'SOAccount', 'SOComm', 'SOPartFreq', 'SOVisitFreq', 'WebframeHaveWorkedWith', 'WebframeWantToWorkWith', 'WorkExp', 'YearsCode']


## Columns common to all four years (2022–2025)

33 columns exist in every survey year. Short explanation of each:

### Identification & filtering
| Column | Meaning | Note |
|---|---|---|
| `ResponseId` | ID number of each respondent | Only unique *within* one year, not across years |
| `MainBranch` | Relation to coding (e.g. developer by profession, learning to code, hobby) | Useful for filtering professional developers |
| `Employment` | Employment status (full-time, part-time, freelancer, student, …) | Can contain several answers separated by `;` – categories changed in 2025 |

### Target
| Column | Meaning | Note |
|---|---|---|
| `ConvertedCompYearly` | Total yearly compensation, converted to US dollars | **Target variable** |
| `CompTotal` | Compensation as entered, in the respondent's own currency | Not comparable across countries – would leak the target |
| `Currency` | Currency used for `CompTotal` | Only describes `CompTotal` – would leak the target |

### Person
| Column | Meaning | Note |
|---|---|---|
| `Age` | Age group (e.g. "25-34 years old") | Categories, not exact numbers |
| `Country` | Country of residence | Many categories – rare countries may need grouping |
| `EdLevel` | Highest level of formal education | Long text labels |
| `LearnCode` | How the person learned to code (books, online, school, …) | Multi-select (`;`-separated) |

### Experience
| Column | Meaning | Note |
|---|---|---|
| `YearsCode` | Years of coding in total (including education) | Text values like "Less than 1 year" in 2022–2024; numeric in 2025 |
| `WorkExp` | Years of professional work experience | Numeric |

### Job
| Column | Meaning | Note |
|---|---|---|
| `DevType` | Job role (e.g. back-end developer, data scientist) | Can contain several roles in some years |
| `OrgSize` | Size of the company (number of employees) | Categories |
| `ICorPM` | Individual contributor or people manager | Two categories |
| `RemoteWork` | Work arrangement: remote, hybrid or in-person | Labels differ between years (e.g. "Fully remote" vs "Remote") |
| `PurchaseInfluence` | How much influence the person has on technology purchases at work | Categories |

### Technology (all multi-select, `;`-separated)
| Column | Meaning |
|---|---|
| `LanguageHaveWorkedWith` | Programming languages used in the past year |
| `LanguageWantToWorkWith` | Programming languages the person wants to use next year |
| `DatabaseHaveWorkedWith` | Databases used |
| `DatabaseWantToWorkWith` | Databases the person wants to use |
| `PlatformHaveWorkedWith` | Cloud platforms used (AWS, Azure, …) |
| `PlatformWantToWorkWith` | Cloud platforms the person wants to use |
| `WebframeHaveWorkedWith` | Web frameworks used (React, Django, …) |
| `WebframeWantToWorkWith` | Web frameworks the person wants to use |
| `OfficeStackAsyncHaveWorkedWith` | Asynchronous collaboration tools used (Jira, Confluence, …) |
| `OfficeStackAsyncWantToWorkWith` | Asynchronous collaboration tools the person wants to use |
| `OpSysPersonal use` | Operating systems used privately |
| `OpSysProfessional use` | Operating systems used at work |

### Stack Overflow usage
| Column | Meaning |
|---|---|
| `SOAccount` | Whether the person has a Stack Overflow account |
| `SOVisitFreq` | How often they visit Stack Overflow |
| `SOPartFreq` | How often they participate (ask/answer) |
| `SOComm` | Whether they feel part of the Stack Overflow community |

### Observations
- `YearsCodePro` (years of professional coding) exists in 2022–2024 but **not in 2025** → `WorkExp` is the common experience column.
- Several columns have **different answer categories** in different years (e.g. `RemoteWork`, `Employment`) → they need harmonising before combining the years.

**Finding Null/NA values**

In [23]:
na_2022 = df_2022[list(common_cols)].isna().mean()*100

In [24]:
na_2023 = df_2023[list(common_cols)].isna().mean()*100

In [25]:
na_2024 = df_2024[list(common_cols)].isna().mean()*100

In [26]:
na_2025 = df_2025[list(common_cols)].isna().mean()*100

In [27]:
na_tabel = pd.DataFrame({"2022": na_2022, "2023":na_2023, "2024":na_2024, "2025":na_2025})
na_tabel.round(1).sort_values(by="2025", ascending = False)

,2022,2023,2024,2025
WebframeWantToWorkWith,37.1,36.4,41.1,64.1
PlatformWantToWorkWith,44.8,42.5,47.2,60.5
DatabaseWantToWorkWith,30.4,31.7,35.0,59.9
WebframeHaveWorkedWith,26.9,24.9,31.0,53.3
OfficeStackAsyncWantToWorkWith,56.2,39.7,40.5,52.8
ConvertedCompYearly,48.0,46.2,64.2,51.3
PlatformHaveWorkedWith,31.9,28.7,35.3,50.7
CompTotal,47.6,45.9,48.4,49.5
DatabaseHaveWorkedWith,17.9,17.7,23.2,48.1
LanguageWantToWorkWith,8.5,9.5,14.8,45.0


### Observations on missing values

- **Salary (`ConvertedCompYearly`) is missing for about half of the respondents** (46–51% in 2022, 2023 and 2025, and 64% in 2024). These rows cannot be used for salary analysis, so they should be removed first. After that, the missing values of the other columns should be checked again.
- **`WorkExp` and `ICorPM` are about 50% empty in 2022–2024, but much less in 2025** (13% and 32%). In the earlier surveys these questions were part of a later section that only about half of the people answered. In 2025, `WorkExp` was asked at the beginning of the survey.
- **In 2025, basic columns have many more missing values than before**, for example `Country` (28% vs. about 2% in 2022) and `LanguageHaveWorkedWith` (36% vs. about 3%). The 2025 survey was much longer (172 columns), so many people probably stopped answering halfway.
- **Empty technology columns** (e.g. `PlatformHaveWorkedWith`) can mean either "skipped the question" or "does not use this technology", which matters when interpreting them.
- **In 2024, about 10,000 people have `CompTotal` but no `ConvertedCompYearly`**, which means their salary was not converted. This should be checked during salary cleaning.

**Combine DSs and Adding year columns**

In [28]:
small_2022 = df_2022[cols].copy()
small_2022["Year"] = 2022

small_2023 = df_2023[cols].copy()
small_2023["Year"] = 2023

small_2024 = df_2024[cols].copy()
small_2024["Year"] = 2024

small_2025 = df_2025[cols].copy()
small_2025["Year"] = 2025

In [29]:
Combined = pd.concat([small_2022, small_2023, small_2024, small_2025], ignore_index = True)

In [30]:
Combined.shape

(277080, 34)

In [31]:
Combined

,Age,CompTotal,ConvertedCompYearly,Country,Currency,DatabaseHaveWorkedWith,DatabaseWantToWorkWith,DevType,EdLevel,Employment,...,ResponseId,SOAccount,SOComm,SOPartFreq,SOVisitFreq,WebframeHaveWorkedWith,WebframeWantToWorkWith,WorkExp,YearsCode,Year
0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2022
1,NaN,NaN,NaN,Canada,CAD\tCanadian dollar,NaN,NaN,NaN,NaN,"Employed, full-time",...,2,Yes,Not sure,Daily or almost daily,Daily or almost daily,NaN,NaN,NaN,NaN,2022
2,25-34 years old,32000.0,40205.0,United Kingdom of Great Britain and Northern I...,GBP\tPound sterling,Microsoft SQL Server,Microsoft SQL Server,Data scientist or machine learning specialist;...,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)","Employed, full-time",...,3,Yes,Neutral,Multiple times per day,Multiple times per day,Angular.js,Angular;Angular.js,NaN,14,2022
3,35-44 years old,60000.0,215232.0,Israel,ILS\tIsraeli new shekel,Microsoft SQL Server,Microsoft SQL Server,"Developer, full-stack","Bachelor’s degree (B.A., B.S., B.Eng., etc.)","Employed, full-time",...,4,Yes,"Yes, definitely",A few times per week,Daily or almost daily,ASP.NET;ASP.NET Core,ASP.NET;ASP.NET Core,NaN,20,2022
4,25-34 years old,NaN,NaN,United States of America,USD\tUnited States dollar,Cloud Firestore;Elasticsearch;Microsoft SQL Se...,Cloud Firestore;Elasticsearch;Firebase Realtim...,"Developer, front-end;Developer, full-stack;Dev...","Bachelor’s degree (B.A., B.S., B.Eng., etc.)","Employed, full-time",...,5,Yes,"Yes, definitely",Daily or almost daily,Multiple times per day,Angular;ASP.NET;ASP.NET Core ;jQuery;Node.js,Angular;ASP.NET Core ;Blazor;Node.js,NaN,8,2022
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
277075,18-24 years old,NaN,NaN,NaN,NaN,NaN,NaN,NaN,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)",Employed,...,49187,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2025
277076,45-54 years old,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Some college/university study without earning ...,Employed,...,49188,NaN,NaN,NaN,NaN,NaN,NaN,25.0,NaN,2025
277077,35-44 years old,80000.0,NaN,Italy,EUR European Euro,NaN,NaN,"Architect, software or solutions","Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",Employed,...,49189,NaN,NaN,NaN,NaN,NaN,NaN,17.0,31.0,2025
277078,25-34 years old,NaN,NaN,NaN,NaN,NaN,NaN,NaN,"Professional degree (JD, MD, Ph.D, Ed.D, etc.)","Independent contractor, freelancer, or self-em...",...,49190,NaN,NaN,NaN,NaN,NaN,NaN,2.0,NaN,2025


In [32]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [33]:
Combined.to_csv('/content/drive/MyDrive/Data_Circle_StackOverflow/combined_2022_2025.csv', index=False)